GraphicscContextCG uses memory redundantly for SourceBrush

kkinnunen-apple opened the pull request:
#### d08bfd77fc2346828b9fc49a9dd6e8a8bbc8e320
<pre>
GraphicscContextCG uses memory redundantly for SourceBrush
<a href="https://bugs.webkit.org/show_bug.cgi?id=267405">https://bugs.webkit.org/show_bug.cgi?id=267405</a>
<a href="https://rdar.apple.com/120841202">rdar://120841202</a>

Reviewed by Said Abou-Hallawa.

SourceBrush would contain std::optional&lt;std::variant&lt;...&gt;&gt;.
This wastes quite a bit of memory to padding and redundant tag field.
Instead, store it just as variant&lt;monostate, ...&gt;
Also store the SourceBrush instances as the first instances,
as SourceBrush has alignof == 8 and ChangeFlags has sizeof == 4.

Before:
Total byte size: 800
Total pad bytes: 60
Padding percentage: 7.50 %

After:
Total byte size: 752
Total pad bytes: 62
Padding percentage: 8.24 %

Some, likely leaky, pages might create thousands of layers or 2d
contexts. GraphicscContextCG for these ImageBuffers end up taking
significant amount of memory, risking GPUP jetsams.

* Source/WebCore/platform/graphics/GraphicsContextState.h:
* Source/WebCore/platform/graphics/SourceBrush.cpp:
(WebCore::SourceBrush::SourceBrush):
(WebCore::SourceBrush::gradientSpaceTransform const):
(WebCore::SourceBrush::gradient const):
(WebCore::SourceBrush::gradientIdentifier const):
(WebCore::SourceBrush::pattern const):
(WebCore::SourceBrush::setGradient):
(WebCore::SourceBrush::setPattern):
* Source/WebCore/platform/graphics/SourceBrush.h:
(WebCore::SourceBrush::SourceBrush):
(WebCore::SourceBrush::patternGradient const):
(WebCore::SourceBrush::isInlineColor const):
(WebCore::SourceBrush::isVisible const):
(WebCore::SourceBrush::hasPatternOrGradient const):
(WebCore::operator==):
(WebCore::SourceBrush::brush const): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/273073@main">https://commits.webkit.org/273073@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76272109da996e938fc92915afeb467525c03f6b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34044 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12839 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36015 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36682 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30856 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35117 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15224 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9978 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29906 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34546 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10865 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30382 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9475 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9587 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30402 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37990 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30931 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30725 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35689 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9711 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7637 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33592 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11496 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7868 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10275 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10522 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/SourceBrush.h:
```diff
@@ -34,22 +34,15 @@ namespace WebCore {
     void setColor(const Color color) { m_color = color; }
 
-    const std::optional<Brush>& brush() const { return m_brush; }
+    const OptionalPatternGradient& patternGradient() const { return m_brush; }
```
This function is only used by the encoder. The rest of the code is either getting the gradient or the pattern out of `m_brush`. Maybe if we add something like this the code will be a little bit cleaner:
```
const Pattern* pattern() const;
const SourceBrushLogicalGradient* gradient() const;
```

kkinnunen-apple replied:
They are there, just below this line.

shallawa commented on Source/WebCore/platform/graphics/SourceBrush.h:
```diff
@@ -59,30 +52,30 @@ class SourceBrush {
 private:
     Color m_color { Color::black };
-    std::optional<Brush> m_brush;
+    OptionalPatternGradient m_brush;
```
Please change the type name or the member name so they match:
```
OptionalBrush m_brush;
OptionalPatternGradient m_patternGradient;
```

shallawa commented on Source/WebCore/platform/graphics/SourceBrush.h:
```diff
@@ -59,30 +52,30 @@ class SourceBrush {
+            return aPattern->ptr() == std::get<Ref<Pattern>>(b).ptr();
+        return true;
+    };
+    return a.m_color == b.m_color && patternGradientEqual(a.m_brush, b.m_brush);
```
Can't this be 
```
return a.m_color == b.m_color && a.gradient() == b.gradient() && a.pattern() == b.pattern();
```

kkinnunen-apple replied:
No, because gradient() is different to SourceBrushLogicalGradient.
E.g. <-- the old implementation compares logical gradient by value and pattern by ptr. For whatever reason..

shallawa commented on Source/WebCore/platform/graphics/SourceBrush.h:
```diff
@@ -34,22 +34,15 @@ namespace WebCore {
-        using Variant = std::variant<LogicalGradient, Ref<Pattern>>;
-        Variant brush;
-    };
+    using OptionalPatternGradient = std::variant<std::monostate, SourceBrushLogicalGradient, Ref<Pattern>>;
```
Nice. This simplifies the definition.

kkinnunen-apple committed:
GraphicscContextCG uses memory redundantly for SourceBrush
https://bugs.webkit.org/show_bug.cgi?id=267405
rdar://120841202

Reviewed by Said Abou-Hallawa.

SourceBrush would contain std::optional<std::variant<...>>.
This wastes quite a bit of memory to padding and redundant tag field.
Instead, store it just as variant<monostate, ...>
Also store the SourceBrush instances as the first instances,
as SourceBrush has alignof == 8 and ChangeFlags has sizeof == 4.

Before:
Total byte size: 800
Total pad bytes: 60
Padding percentage: 7.50 %

After:
Total byte size: 752
Total pad bytes: 62
Padding percentage: 8.24 %

Some, likely leaky, pages might create thousands of layers or 2d
contexts. GraphicscContextCG for these ImageBuffers end up taking
significant amount of memory, risking GPUP jetsams.

* Source/WebCore/platform/graphics/GraphicsContextState.h:
* Source/WebCore/platform/graphics/SourceBrush.cpp:
(WebCore::SourceBrush::SourceBrush):
(WebCore::SourceBrush::gradientSpaceTransform const):
(WebCore::SourceBrush::gradient const):
(WebCore::SourceBrush::gradientIdentifier const):
(WebCore::SourceBrush::pattern const):
(WebCore::SourceBrush::setGradient):
(WebCore::SourceBrush::setPattern):
* Source/WebCore/platform/graphics/SourceBrush.h:
(WebCore::SourceBrush::SourceBrush):
(WebCore::SourceBrush::patternGradient const):
(WebCore::SourceBrush::isInlineColor const):
(WebCore::SourceBrush::isVisible const):
(WebCore::SourceBrush::hasPatternOrGradient const):
(WebCore::operator==):
(WebCore::SourceBrush::brush const): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/273073@main
